### PR TITLE
Fix SSR 'case B:b' with universe polymorphic equality

### LINKED ISSUE #10301: SSR case tactic with universe polymorphism
#### Description of the problem
In the following code the `case B:b` fails with `Illegal application...`
```
Require Import ssreflect.

Polymorphic Cumulative Variant paths {A} (x:A) : A -> Type
  := idpath : paths x x.

Register paths as core.eq.type.
Register idpath as core.eq.refl.

Lemma case_test (b:bool) : paths b b.
Proof. case B:b; reflexivity. Qed.
```
The problem is in `ssrelim.ml`, where it is assumed that `core.eq.type` lives in `Prop`, which is not the case in HoTT.

#### Coq Version

8.10+alpha

### PR DESCRIPTION
**Pull request by andreaslyn**
**Kind:** bug fix.

Fixes / closes #10301

- [x] Added / updated test-suite

- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [ ] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: plugins/ssr/ssrelim.ml [hunk 1]
```diff
@@ -383,15 +383,22 @@ let ssrelim ?(is_case=false) deps what ?elim eqid elim_intro_tac =
...
+          let eq_ty = Retyping.get_type_of (pf_env gl) (project gl) erefl_ty in
+          let gen_eq_tac s =
+            let open Evd in
+            let sigma = merge_universe_context s.sigma (evar_universe_context (project gl)) in
+            apply_type new_concl [erefl] { s with sigma }
+          in
+          gen_eq_tac, eq_ty, gl in
         let rel = k + if c_is_head_p then 1 else 0 in
-        let src, gl = mkProt EConstr.mkProp EConstr.(mkApp (eq,[|t; c; mkRel rel|])) gl in
+        let src, gl = mkProt eq_ty EConstr.(mkApp (eq,[|t; c; mkRel rel|])) gl in
```
Comment (andreaslyn) [inline plugins/ssr/ssrelim.ml]: Replace `Prop` with the type of the equality.

File: plugins/ssr/ssrelim.ml [hunk 2]
```diff
@@ -383,15 +383,22 @@ let ssrelim ?(is_case=false) deps what ?elim eqid elim_intro_tac =
...
           let new_concl = EConstr.mkArrow refl Sorts.Relevant (EConstr.Vars.lift 1 (pf_concl orig_gl)) in
           let new_concl = fire_subst gl new_concl in
           let erefl, gl = mkRefl t c gl in
           let erefl = fire_subst gl erefl in
-          apply_type new_concl [erefl], gl in
+          let erefl_ty = Retyping.get_type_of (pf_env gl) (project gl) erefl in
+          let eq_ty = Retyping.get_type_of (pf_env gl) (project gl) erefl_ty in
+          let gen_eq_tac s =
+            let open Evd in
+            let sigma = merge_universe_context s.sigma (evar_universe_context (project gl)) in
```
Comment (andreaslyn) [inline plugins/ssr/ssrelim.ml]: Add the universes from the equality type to the context.

**Review by gares — APPROVED** [review]
Looks good. Thanks!

**Comment by gares** [comment]
I've started a bench, it will be over by tomorrow
 https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/727/console
I imagine no problems, but since we typecheck a bit more I'd like to know if there is any impact on performances.

**Comment by gares** [comment]
```
┌────────────────────────┬─────────────────────────┬───────────────────────────────────────────┬───────────────────────────────────────────┬───────────────────────────────┬──────────────────────────┐
│                        │      user time [s]      │                CPU cycles                 │             CPU instructions              │     max resident mem [KB]     │        mem faults        │
│                        │                         │                                           │                                           │                               │                          │
│           package_name │     NEW     OLD PDIFF   │              NEW              OLD PDIFF   │              NEW              OLD PDIFF   │        NEW        OLD PDIFF   │    NEW    OLD    PDIFF   │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│            coq-bignums │   69.01   69.23 -0.32 % │     192458586282     193079394606 -0.32 % │     253769273733     253708426273 +0.02 % │     491176     491252 -0.02 % │    164    135   +21.48 % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-character │  187.82  188.17 -0.19 % │     523068573229     523421962637 -0.07 % │     693642905032     693602770545 +0.01 % │    1104340    1101940 +0.22 % │      6      0     +nan % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│     coq-mathcomp-field │  230.94  231.02 -0.03 % │     643436794456     642902009487 +0.08 % │     918466669471     918484148275 -0.00 % │     755864     756048 -0.02 % │      4      0     +nan % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-odd-order │ 1462.26 1462.04 +0.02 % │    4073904984117    4075359129527 -0.04 % │    6974959278277    6974551066666 +0.01 % │    1267296    1267428 -0.01 % │      1      0     +nan % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│               coq-hott │  334.04  333.40 +0.19 % │     906925313072     906487221988 +0.05 % │    1376623719812    1376575018536 +0.00 % │     477744     477924 -0.04 % │    442      0     +nan % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-ssreflect │   40.01   39.93 +0.20 % │     110627721716     110482689221 +0.13 % │     133668602506     133635181996 +0.03 % │     528444     528476 -0.01 % │     11     55   -80.00 % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-solvable │  190.30  189.82 +0.25 % │     530501200568     529481251322 +0.19 % │     707259682386     707309848421 -0.01 % │     781016     823404 -5.15 % │     12      2  +500.00 % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│   coq-mathcomp-algebra │  166.88  166.41 +0.28 % │     465069098815     464772072334 +0.06 % │     574151570391     574104420643 +0.01 % │     618100     618196 -0.02 % │     85      3 +2733.33 % │
├────────────────────────┼─────────────────────────┼───────────────────────────────────────────┼───────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-fingroup │   51.74   51.56 +0.35 % │     144048325567     143805633125 +0.17 % │     187144711337     187138309648 +0.00 % │     567960     567924 +0.01 % │      1     38   -97.37 % │
└────────────────────────┴─────────────────────────┴───────────────────────────────────────────┴───────────────────────────────────────────┴───────────────────────────────┴──────────────────────────┘
```
Awesome, the bench shows just noise. I can merge

**Comment by gares** [comment]
I don't think you need to add anything to changes, but you can add your name to CREDITS if you like.

**Comment by gares** [comment]
Good, I'll merge tomorrow

**Pull request merged**